src/main/java/net/imagej/ops/OpSignature.java

```diff
@@ -0,0 +1,188 @@
+ * The minimum required components for a call to
+ * {@link OpService#run(String, Object...)}
+ *
+ * @author Gabriel Selzer
+ */
+public class OpSignature {
+
+	private final String name;
+	private final String functionalType;
+	private final List<ModuleItem<?>> inputs;
+	private final List<ModuleItem<?>> outputs;
+
+	public OpSignature(final OpInfo info) {
+		this(info.getName(), info.getType(), info.inputs(), info.outputs());
+	}
+
+	public OpSignature(final String name, final Class<?> functionalType,
+		final List<ModuleItem<?>> inputs, final List<ModuleItem<?>> outputs)
+	{
+		this.name = name;
+		this.functionalType = functionalType.getName();
+		this.inputs = inputs;
+		this.outputs = outputs;
+	}
+
+	public String getName() {
+		return name;
+	}
+
+	public String getFunctionalType() {
+		return functionalType;
+	}
+
+	public List<ModuleItem<?>> inputsFor(ModuleInfo info) {
+		return inputs.stream() //
+			.map(item -> new DefaultMutableModuleItem<>(info, item)) //
+			.collect(Collectors.toList());
+	}
+
+	public List<Type> inputTypes() {
+		return inputs.stream() //
+			.map(ModuleItem::getGenericType) //
+			.collect(Collectors.toList());
+	}
+
+	public List<String> inputNames() {
+		return inputs.stream() //
+			.map(ModuleItem::getName) //
+			.collect(Collectors.toList());
+	}
+
+	public List<ModuleItem<?>> outputsFor(ModuleInfo info) {
+		return outputs.stream() //
+			.map(item -> new DefaultMutableModuleItem<>(info, item)) //
+			.collect(Collectors.toList());
+	}
+
+	public List<Type> outputTypes() {
+		return outputs.stream() //
+			.map(ModuleItem::getGenericType) //
+			.collect(Collectors.toList());
+	}
+
+	public List<String> outputNames() {
+		return outputs.stream() //
+			.map(ModuleItem::getName) //
+			.collect(Collectors.toList());
+	}
+
+	@Override
+	public boolean equals(Object that) {
+		// ensure that is a OpSignatureInfo
+		if (!(that instanceof OpSignature)) return false;
+		OpSignature thatSig = (OpSignature) that;
+		// check name
+		return buildShortSig().equals(thatSig.buildShortSig());
+	}
+
+	@Override
+	public int hashCode() {
+		return Objects.hash(buildShortSig());
+	}
+
+	@Override
+	public String toString() {
+		return buildShortSig();
+	}
+
+	// -- Helper methods --
+
+	public String buildShortSig() {
+		final StringBuilder sb = new StringBuilder();
+		// name
+		sb.append(getName());
+		// inputs
+		sb.append("(").append(paramList(inputs, "")).append(")");
+		// (pure) outputs
+		List<ModuleItem<?>> pureOutputs = outputs.stream() //
+			.filter(item -> !item.isInput()) //
+			.collect(Collectors.toList());
+		sb.append(paramList(pureOutputs, " -> "));
+		return sb.toString();
+	}
+
+	private String paramList(Iterable<ModuleItem<?>> items, String beforeFirst) {
+		boolean first = true;
+		StringBuilder sb = new StringBuilder();
+		for (ModuleItem<?> item : items) {
+			if (first) {
+				sb.append(beforeFirst);
+				first = false;
+			}
+			else sb.append(", ");
+			sb.append(shortParam(item, true));
```

Comment:
Remove `true`, just hardcode that variable in `shortParam`.

We can probably simplify `shortParam` if we do this.

Reply:
Fixed.